File: frontend/src/lib/api/metriche.ts
/**
 * Fetcher per `/metriche` e `/utenti/{id}/metriche` sul backend FastAPI
 * (issue #7, PRD entità "Metrica di lettura"). Stesse convenzioni di
 * `lib/api/voci.ts`: unioni discriminate per ogni esito, mapping
 * snake_case -> camelCase a carico di questo modulo, mai un'eccezione
 * per il flusso di controllo.
 */

import { intestazioniConLingua } from "@/lib/lingua";

export type VoceClassifica = {
  id: string;
  nome: string;
  /** Frazionario quando il Libro ha più di un autore/genere (PRD regola
   * 18): 1,5 per un autore su un libro a due autori, non arrotondato a
   * un intero (design-frontend.md §14). */
  peso: number;
};

export type Metriche = {
  anno: number;
  /** Intervallo selezionabile (PRD, comportamento #12): dal primo anno
   * con dati all'anno corrente. Nessun dato -> i due coincidono
   * sull'anno corrente. */
  annoMinimo: number;
  annoMassimo: number;
  libriFiniti: number;
  /** Quante Letture concluse nell'anno sono una rilettura dello stesso
   * Libro: "di cui N riletture" (design-frontend.md §14). */
  riletture: number;
  /** Somma degli incrementi datati nell'anno, mai delle pagine
   * raggiunte (PRD): include anche le Letture non concluse o
   * abbandonate. */
  pagineLette: number;
  autoriPiuLetti: VoceClassifica[];
  generiPrincipali: VoceClassifica[];
  /** Libri finiti nell'anno senza alcun genere assegnato: lo scarto da
   * dichiarare accanto a `generiPrincipali` (design-frontend.md §14). */
  libriSenzaGenere: number;
  /** Governa la spiegazione della divergenza a cavallo d'anno, mostrata
   * solo quando serve (design-frontend.md §14). */
  haLettureACavalloAnno: boolean;
  /** Quante sono: `haLettureACavalloAnno` decide SE mostrare la
   * spiegazione, questo le permette di andare al plurale quando serve.
   * Prima la frase era al singolare fisso e con due letture mentiva. */
  lettureACavalloAnno: number;

  // --- Aggiunte dal ridisegno degli Annali (design-frontend.md §14) ------

  /** Dodici numeri, da gennaio a dicembre, sempre di lunghezza 12 anche
   * per un anno in corso. È `pagineLette` a una risoluzione più fine:
   * la somma delle caselle coincide col totale per costruzione. */
  paginePerMese: number[];
  /** Date distinte con almeno un avanzamento a incremento positivo:
   * misura l'abitudine, non il volume. */
  giorniConLettura: number;
  /** Giorni dell'anno già trascorsi: l'anno intero se è passato, il
   * giorno dell'anno se è quello in corso. Denominatore obbligato di
   * `giorniConLettura`, che da solo non si può leggere. */
  giorniTrascorsi: number;
  /** Media dei voti dei libri finiti nell'anno che un voto ce l'hanno.
   * `null` quando nessuno è votato, mai 0 (che sarebbe un voto pessimo
   * invece di un'assenza). */
  votoMedio: number | null;
  /** Il denominatore esplicito di `votoMedio`. */
  libriVotati: number;
  /** Cinque caselle, da una a cinque stelle. I mezzi voti si
   * arrotondano alla stella superiore. */
  votiPerStella: number[];
  /** Letture abbandonate chiuse nell'anno: fuori da `libriFiniti`
   * (regola 13), ma le loro pagine restano dentro `pagineLette`. */
  abbandoni: number;
  /** Estremi inclusi: una lettura aperta e chiusa in giornata dura un
   * giorno. `null` quando nell'anno non si è conclusa alcuna lettura. */
  durataMediaGiorni: number | null;
  durataMassimaGiorni: number | null;
  durataMassimaTitolo: string | null;
  /** Libri finiti senza pagine adottate: lo scarto che rende concreto
   * il limite su `pagineLette`. Zero significa somma completa, e allora
   * il limite non va nemmeno scritto. */
  libriSenzaPagine: number;
};

type VoceClassificaBody = { id: string; nome: string; peso: number };

type MetricheBody = {
  anno: number;
  anno_minimo: number;
  anno_massimo: number;
  libri_finiti: number;
  riletture: number;
  pagine_lette: number;
  autori_piu_letti: VoceClassificaBody[];
  generi_principali: VoceClassificaBody[];
  libri_senza_genere: number;
  ha_letture_a_cavallo_anno: boolean;
  letture_a_cavallo_anno: number;
  pagine_per_mese: number[];
  giorni_con_lettura: number;
  giorni_trascorsi: number;
  voto_medio: number | null;
  libri_votati: number;
  voti_per_stella: number[];
  abbandoni: number;
  durata_media_giorni: number | null;
  durata_massima_giorni: number | null;
  durata_massima_titolo: string | null;
  libri_senza_pagine: number;
};

function toMetriche(body: MetricheBody): Metriche {
  return {
    anno: body.anno,
    annoMinimo: body.anno_minimo,
    annoMassimo: body.anno_massimo,
    libriFiniti: body.libri_finiti,
    riletture: body.riletture,
    pagineLette: body.pagine_lette,
    autoriPiuLetti: body.autori_piu_letti,
    generiPrincipali: body.generi_principali,
    libriSenzaGenere: body.libri_senza_genere,
    haLettureACavalloAnno: body.ha_letture_a_cavallo_anno,
    lettureACavalloAnno: body.letture_a_cavallo_anno,
    paginePerMese: body.pagine_per_mese,
    giorniConLettura: body.giorni_con_lettura,
    giorniTrascorsi: body.giorni_trascorsi,
    votoMedio: body.voto_medio,
    libriVotati: body.libri_votati,
    votiPerStella: body.voti_per_stella,
    abbandoni: body.abbandoni,
    durataMediaGiorni: body.durata_media_giorni,
    durataMassimaGiorni: body.durata_massima_giorni,
    durataMassimaTitolo: body.durata_massima_titolo,
    libriSenzaPagine: body.libri_senza_pagine,
  };
}

function baseUrlOrError(): { baseUrl: string } | { status: "error"; message: string } {
  const baseUrl = process.env.NEXT_PUBLIC_API_BASE_URL;
  if (!baseUrl) {
    return { status: "error", message: "L’app non è configurata come dovrebbe. Parla con chi mantiene l’istanza." };
  }
  return { baseUrl };
}

export type MetricheResult =
  | { status: "ok"; data: Metriche }
  // L'anno richiesto è oltre l'anno corrente (PRD: "gli anni futuri non
  // sono selezionabili").
  | { status: "anno_futuro" }
  | { status: "error"; message: string };

/**
 * GET /metriche: le proprie metriche di lettura. `anno` omesso ->
 * l'anno corrente in Europa centrale, deciso dal backend.
 *
 * `acceptLanguage`, opzionale (issue #34): inoltra la lingua dell'interfaccia
 * al backend, che sceglie con la stessa intestazione l'etichetta di genere
 * nella lingua giusta (`app/core/lingua.py`) — vedi il docstring di
 * `lib/api/voci.ts::getVoci`, stesso meccanismo.
 */
export async function getMetriche(
  accessToken: string,
  anno?: number,
  acceptLanguage?: string,
): Promise<MetricheResult> {
  const config = baseUrlOrError();
  if ("status" in config) return config;

  const query = anno !== undefined ? `?anno=${anno}` : "";

  let response: Response;
  try {
    response = await fetch(`${config.baseUrl}/metriche${query}`, {
      headers: intestazioniConLingua(accessToken, acceptLanguage),
      cache: "no-store",
    });
  } catch {
    return { status: "error", message: "Il server non risponde. Controlla la connessione e riprova." };
  }

  if (response.status === 422) {
    return { status: "anno_futuro" };
  }
  if (!response.ok) {
    return { status: "error", message: "Il server ha risposto male. Riprova fra poco." };
  }

  const body = (await response.json()) as MetricheBody;
  return { status: "ok", data: toMetriche(body) };
}

export type MetricheCollegatoResult =
  | { status: "ok"; data: Metriche }
  | { status: "anno_futuro" }
  // L'utente indicato non esiste.
  | { status: "not_found" }
  // Nessun collegamento attivo: le sue metriche non sono (più)
  // accessibili — stesso trattamento di `getLibreriaCollegato`.
  | { status: "non_collegato" }
  | { status: "error"; message: string };

/** GET /utenti/{id}/metriche: le metriche di un collegato, stesso
 * payload delle proprie. `acceptLanguage`, opzionale: vedi `getMetriche`. */
export async function getMetricheCollegato(
  accessToken: string,
  utenteId: string,
  anno?: number,
  acceptLanguage?: string,
): Promise<MetricheCollegatoResult> {
  const config = baseUrlOrError();
  if ("status" in config) return config;

  const query = anno !== undefined ? `?anno=${anno}` : "";

  let response: Response;
  try {
    response = await fetch(`${config.baseUrl}/utenti/${utenteId}/metriche${query}`, {
      headers: intestazioniConLingua(accessToken, acceptLanguage),
      cache: "no-store",
    });
  } catch {
    return { status: "error", message: "Il server non risponde. Controlla la connessione e riprova." };
  }

  if (response.status === 404) {
    return { status: "not_found" };
  }
  if (response.status === 403) {
    return { status: "non_collegato" };
  }
  if (response.status === 422) {
    return { status: "anno_futuro" };
  }
  if (!response.ok) {
    return { status: "error", message: "Il server ha risposto male. Riprova fra poco." };
  }

  const body = (await response.json()) as MetricheBody;
  return { status: "ok", data: toMetriche(body) };
}

File: backend/app/schemas/metriche.py
"""Contratti di `GET /metriche` e `GET /utenti/{id}/metriche` (issue #7,
PRD entità "Metrica di lettura"). Stesso payload per le proprie metriche
e per quelle di un collegato — solo la sorgente dei dati cambia."""

from pydantic import BaseModel


class VoceClassificaResponse(BaseModel):
    """Una riga di "autori più letti" o "generi principali": il peso è
    frazionario quando un Libro ha più di un autore/genere (PRD regola
    18, "la somma dei contributi resta pari a un libro") — design-
    frontend.md §14 impone di mostrare i decimali, non arrotondarli a un
    intero."""

    id: str
    nome: str
    peso: float


class MetricheResponse(BaseModel):
    anno: int
    # Intervallo selezionabile (PRD, comportamento #12): dal primo anno
    # con dati all'anno corrente, estremi inclusi. Nessun dato -> i due
    # coincidono sull'anno corrente.
    anno_minimo: int
    anno_massimo: int
    libri_finiti: int
    # Quante delle Letture concluse nell'anno sono una rilettura dello
    # stesso Libro (PRD: "due riletture concluse nello stesso anno
    # contano due" — l'unità è la Lettura, non il Libro).
    riletture: int
    # Somma degli incrementi datati nell'anno (PRD), mai delle pagine
    # raggiunte: include anche le pagine di Letture non ancora concluse
    # o abbandonate (regola 13).
    pagine_lette: int
    autori_piu_letti: list[VoceClassificaResponse]
    generi_principali: list[VoceClassificaResponse]
    # Libri finiti nell'anno senza alcun genere assegnato ("non
    # classificato", PRD): restano fuori da `generi_principali`, quindi
    # la somma dei pesi lì dentro è inferiore a `libri_finiti` esatta-
    # mente di questo scarto (design-frontend.md §14: "lo scarto è
    # dichiarato accanto").
    libri_senza_genere: int
    # Vero se almeno una Lettura conclusa nell'anno è iniziata l'anno
    # prima: governa la spiegazione della divergenza a cavallo d'anno
    # (design-frontend.md §14), mostrata solo quando serve.
    ha_letture_a_cavallo_anno: bool
    # Quante delle Letture concluse nell'anno erano già cominciate
    # nell'anno precedente. `ha_letture_a_cavallo_anno` resta il flag che
    # decide SE mostrare la spiegazione; questo è il numero che le
    # permette di essere al plurale quando serve, invece del singolare
    # fisso che la frase portava prima.
    letture_a_cavallo_anno: int

    # ------------------------------------------------------------------
    # Aggiunte dal ridisegno degli Annali (design-frontend.md §14).
    # Nessuna migrazione e nessuna query nuova: tutte escono dalle stesse
    # tre letture che il servizio faceva già.
    # ------------------------------------------------------------------

    # Somma degli incrementi per mese, dodici numeri da gennaio a
    # dicembre, sempre di lunghezza 12 anche per un anno in corso (i mesi
    # non ancora arrivati valgono zero, e il frontend li distingue da un
    # mese a zero perché conosce `giorni_trascorsi`). È la stessa somma
    # di `pagine_lette`, solo non collassata: sum(pagine_per_mese) ==
    # pagine_lette per costruzione.
    pagine_per_mese: list[int]
    # Date distinte in cui esiste almeno un Avanzamento con incremento
    # positivo. Misura l'abitudine, non il volume: un Avanzamento a
    # incremento zero (una correzione, una pagina riscritta uguale) non
    # è un giorno di lettura.
    giorni_con_lettura: int
    # Giorni dell'anno già trascorsi al momento della richiesta, nel fuso
    # Europa centrale: l'anno intero per un anno passato, il giorno
    # dell'anno per quello corrente. Senza questo denominatore
    # `giorni_con_lettura` non si può leggere, perché 118 su un anno
    # finito e 118 su otto mesi non dicono la stessa cosa.
    giorni_trascorsi: int
    # Media dei voti delle Voci concluse nell'anno che UN voto ce l'hanno,
    # arrotondata a un decimale; `None` quando nessuna è votata (mai 0,0,
    # che sarebbe un voto pessimo invece di un'assenza).
    voto_medio: float | None
    # Quante Voci concluse nell'anno portano un voto: il denominatore
    # esplicito di `voto_medio` (design-frontend.md §14, "ogni numero
    # porta accanto il suo limite").
    libri_votati: int
    # Distribuzione dei voti in cinque caselle, da una a cinque stelle.
    # I mezzi voti si arrotondano per eccesso alla stella superiore: la
    # distribuzione è un istogramma a cinque colonne, non la scala a
    # dieci passi con cui si vota.
    voti_per_stella: list[int]
    # Letture con esito 'abbandonata' chiuse nell'anno. Non entrano in
    # `libri_finiti` (regola 13) ma le loro pagine sì, come quelle di
    # qualunque altra Lettura: sono pagine che l'Utente ha letto.
    abbandoni: int
    # Durata in giorni di una Lettura conclusa nell'anno, dalla data di
    # inizio a quella di fine, estremi inclusi (una Lettura cominciata e
    # conclusa in giornata dura un giorno, non zero). `None` quando
    # nell'anno non si è conclusa alcuna Lettura.
    durata_media_giorni: int | None
    durata_massima_giorni: int | None
    # Il titolo della Lettura più lunga, nella lingua dell'interfaccia
    # quando esiste una variante. `None` insieme a `durata_massima_giorni`.
    durata_massima_titolo: str | None
    # Libri finiti nell'anno senza `pagine_adottate` sulla Voce: lo scarto
    # che rende concreto il limite su `pagine_lette`. Zero significa che
    # la somma è completa e il limite non va nemmeno scritto.
    libri_senza_pagine: int
